### ISSUE TITLE
Speed up checksum calculation for items

### ISSUE THREAD

**Issue by RubenVerborgh** [nanoc/nanoc#790]
I noticed that roughly 80% of my site's compilation time is due to checksum calculation. While #789 eliminates some of that time, it still weighs on the speed of nanoc.

I think that (some) items can be smarter about their checksum calculation. For instance, an item from the filesystem could generate its checksum based on the SHA1 hash of the file contents, rather than having to parse the entire file and calculating a SHA1 hash of all fields, as is done now. Other data sources could calculate checksums even faster. For instance, the [BibTeX datasource](https://github.com/RubenVerborgh/nanoc-bibtex) could generate one checksum per .bib file, instead of having to generate a checksum for each of the individual entries in that file.

I would like to implement this idea and submit a pull request, but I'd appreciate some guidance on how to implement this. In particular, I'm wondering about the following:

- Currently, `Nanoc::Int::Checksummer` is responsible for its checksumming all other classes. However, this makes that component rather omniscient, in contrast with the principles of object-oriented design. Wouldn't it be good that each entity is responsible for checksumming itself?
  - If so, this would mean we need to extend basic classes like `String` with a checksum method. This might or might not be desired.
  - An alternative is to implement `Checksummer` such that it first checks whether the object supports a `calc_checksum` method and if so, calls it.
- I see two ways in which a datasource could work:
  - Either it would generate a field such as `:checksum`, which the `checksum` method then uses if it exists.
  - Otherwise, it would directly implement the `calc_checksum` method on the item it returns, either by monkey patching or subclassing. However, I wonder to what extent subclassing is desired, given that `Item` has moved to the Nanoc internals.
    - On the theme of subclassing and speed: having subclasses of `Item` would also allow us to have dedicated methods to access properties, which would allow lazy loading. Alternatively, the `new_item` method could accept a block as parameter, which is only executed when the properties are really needed.
- What does a `calc_checksum` look like? Does it take a digest as argument, which it then updates? Or does it take no arguments?

**Comment by ddfreyne** [comment]
> Other data sources could calculate checksums even faster.

That’s an interesting idea! The advantage of the current approach to generating checksums is that no-op changes (e.g. formatting changes) won’t cause a change in the checksum, but I guess that’s really just an edge case.

> Currently, `Nanoc::Int::Checksummer` is responsible for its checksumming all other classes. However, this makes that component rather omniscient, in contrast with the principles of object-oriented design. Wouldn't it be good that each entity is responsible for checksumming itself?

In nanoc 3.x, the individual entities had a `#checksum` method, but I moved all of that into a single class so that there’s only a single place where checksums are calculated, and there are no more monkey-patched `String`, `Array`, … etc classes.

One could indeed argue that this new approach goes against traditional “good practices” of object-orientation, but I find the code of the checksummer far nicer to read.

> Either it would generate a field such as `:checksum`, which the checksum method then uses if it exists.

I like that idea! That’d mean `::Nanoc::Int::Document` gets an optional checksum, and the `#new_item` and `#new_layout` methods gain a `:checksum` param.

I’m not such a big fan of the idea of implementing `#calc_checksum` on items/layouts and giving it the ability to calculate its own checksum. That seems like logic that should live elsewhere. I believe that the `:checksum` param will make lazy loading possible anyway.

**Comment by RubenVerborgh** [comment]
> One could indeed argue that this new approach goes against traditional “good practices” of object-orientation, but I find the code of the checksummer far nicer to read.

Probably a matter of taste, `case` statements based on type always look a little suspicious to me.

> I like that idea! That’d mean ::Nanoc::Int::Document gets an optional checksum, and the #new_item and #new_layout methods gain a :checksum param.

Yes, that would definitely be a gain already! I can make a pull request that implements this—or would you prefer to work on this yourself?

> I believe that the :checksum param will make lazy loading possible anyway.

Not sure. The signature of `new_item` is currently

```
#new_item(raw_content_or_raw_filename, attributes, identifier, params = {})
```

and this requires me to give the attributes as a hash – so I should have the attributes already. (Or can I make the hash lazy?)

For example, in the case of the BibTeX datasource, it would be beneficial to just supply the checksum (which does not require the file to be parsed) and only calculate the attributes when necessary (which requires the file to be parsed).

**Comment by ddfreyne** [comment]
>> One could indeed argue that this new approach goes against traditional “good practices” of object-orientation, but I find the code of the checksummer far nicer to read.

> Probably a matter of taste, `case` statements based on type always look a little suspicious to me.

I understand your point (I typically feel the same), but I nonetheless am happier with the current checksumming method.

>> I believe that the :checksum param will make lazy loading possible anyway.
>
> Not sure. The signature of `new_item` is currently
> 
> ```
> #new_item(raw_content_or_raw_filename, attributes, identifier, params = {})
> ```
> 
> and this requires me to give the attributes as a hash – so I should have the attributes already. (Or can I make the hash lazy?)
>
> For example, in the case of the BibTeX datasource, it would be beneficial to just supply the checksum (which does not require the file to be parsed) and only calculate the attributes when necessary (which requires the file to be parsed).

I’m not referring to including the checksum in the `attributes` hash—the checksum would be separate (passed in `params`, stored separately from attributes). At the moment, attributes are always loaded, but in the future this might no longer be the case.

**Comment by RubenVerborgh** [comment]
> I’m not referring to including the checksum in the attributes hash—the checksum would be separate (passed in params, stored separately from attributes).

Indeed, that was what I understood.

> At the moment, attributes are always loaded, but in the future this might no longer be the case.

Yes, and the checksum param could actually be an enabler for this: only if the checksum is different, the attributes need to be loaded. This could provide a major speed gain.

Alright if I give the checksum param a go in a PR?

**Comment by RubenVerborgh** [comment]
BTW Do you have examples of other `params` use cases?

**Comment by ddfreyne** [comment]
> Alright if I give the checksum param a go in a PR?

Do it! (On the `master` branch please.)

> BTW Do you have examples of other `params` use cases?

There’s `:binary`, which only exists for items, to indicate whether an item is textual or binary. Most (all?) of this code uses Ruby 2.1-style named arguments btw.

**Comment by RubenVerborgh** [comment]
Implemented in #793.
